src/main/scala/is/hail/methods/Skat.scala

```diff
@@ -0,0 +1,223 @@
+    val (q, pVal) = SPG.computeSkatStats()
+    SkatStat(q, pVal)
+  }
+}
+
+case class SkatAgg[T <: Vector[Double]](q: Double, xws: ArrayBuilder[T], qtxws: ArrayBuilder[DenseVector[Double]])
+
+object Skat {
+  def apply(vds: VariantDataset,
+    keyName: String,
+    variantKeys: String,
+    singleKey: Boolean,
+    weightExpr: String,
+    yExpr: String,
+    covExpr: Array[String], hardCalls: Boolean): KeyTable = {
+
+    if (hardCalls)
+      polymorphicSkat(vds, keyName, variantKeys, singleKey, weightExpr, yExpr, covExpr,
+                      RegressionUtils.hardCalls(_,_), SkatAgg.zeroValSparse, SkatAgg.sparseResultOp _)
+    else
+      {val dosages = (gs:Iterable[Genotype],n:Int) => RegressionUtils.dosages(gs, (0 until n).toArray)
+      polymorphicSkat(vds, keyName, variantKeys, singleKey, weightExpr, yExpr, covExpr,
+                      dosages, SkatAgg.zeroValDense, SkatAgg.denseResultOp _)}
+  }
+
+  def polymorphicSkat[T <: Vector[Double]](vds: VariantDataset,
+    keyName: String,
+    variantKeys: String,
+    singleKey: Boolean,
+    weightExpr: String,
+    yExpr: String,
+    covExpr: Array[String],
+    getGenotype: (Iterable[Genotype], Int) => T, zero: SkatAgg[T],
+    resultOp:(SkatAgg[T], Double) => SkatStat): KeyTable =
+  {
+
+    //get variables
+    val (y, cov, completeSamples) = RegressionUtils.getPhenoCovCompleteSamples(vds, yExpr, covExpr)
+
+    val n = y.size
+    val k = cov.cols
+    val d = n - k
+
+    if (d < 1)
+      fatal(s"$n samples and $k ${ plural(k, "covariate") } including intercept implies $d degrees of freedom.")
+
+    // fit null model
+    val qr.QR(q, r) = qr.reduced.impl_reduced_DM_Double(cov)
+    val beta = r \ (q.t * y)
+    val res = y - cov * beta
+    val sigmaSq = (res dot res) / d
+    val filteredVds = vds.filterSamplesList(completeSamples.toSet)
+
+    val (keysType, keysQuerier) = filteredVds.queryVA(variantKeys)
+    val (weightType, weightQuerier) = filteredVds.queryVA(weightExpr)
+
+    // FIXME update to use Tnumeric class?
+    weightType match {
+      case TDouble => weightQuerier.asInstanceOf[Annotation => Double]
+      case TInt => weightQuerier.asInstanceOf[Annotation => Double]
+      case TLong => weightQuerier.asInstanceOf[Annotation => Double]
+      case TFloat => weightQuerier.asInstanceOf[Annotation => Double]
+      case TDouble => weightQuerier.asInstanceOf[Annotation => Double]
+      case TBoolean => weightQuerier.asInstanceOf[Annotation => Double]
+      case _ => fatal("Weights must be numerical")
+    }
+
+    val sc = filteredVds.sparkContext
+    val resBc = sc.broadcast(res)
+
+    val (keyType, keyedRdd) =
+      if (singleKey) {
+        (keysType, filteredVds.rdd.flatMap { case (v, (va, gs)) =>
+          (Option(keysQuerier(va)), Option(weightQuerier(va))) match {
+            case (Some(key), Some(w)) =>
+              val sqrtw = math.sqrt(w.asInstanceOf[Double])
```

Comment:
when you say boolean weight, do you explicitly mean "True/False" or 0's and 1's. Because if it's the former, it definitely throws an error.